drivers/bridge/bridge.go

```diff
@@ -615,92 +581,114 @@ func (d *driver) CreateNetwork(id string, option map[string]interface{}, nInfo d
 }
 
 func (d *driver) createNetwork(config *networkConfiguration) error {
-	var err error
-
 	defer osl.InitOSContext()()
 
-	networkList := d.getNetworks()
-	for i, nw := range networkList {
-		nw.Lock()
-		nwConfig := nw.config
-		nw.Unlock()
-		if err := nwConfig.Conflicts(config); err != nil {
-			if config.DefaultBridge {
-				// We encountered and identified a stale default network
-				// We must delete it as libnetwork is the source of thruth
-				// The default network being created must be the only one
-				// This can happen only from docker 1.12 on ward
-				logrus.Infof("Removing stale default bridge network %s (%s)", nwConfig.ID, nwConfig.BridgeName)
-				if err := d.DeleteNetwork(nwConfig.ID); err != nil {
-					logrus.Warnf("Failed to remove stale default network: %s (%s): %v. Will remove from store.", nwConfig.ID, nwConfig.BridgeName, err)
-					d.storeDelete(nwConfig)
-				}
-				networkList = append(networkList[:i], networkList[i+1:]...)
-			} else {
+	// The next critical section guarantees that a new network is inserted
+	// in the map only if it passes the conflict check with the current configuration
+	d.Lock()
+	for nwID, nwObj := range d.networks {
+		// Config validation
+		if err := nwObj.config.Conflicts(config); err != nil {
+			if !config.DefaultBridge {
+				d.Unlock()
 				return types.ForbiddenErrorf("cannot create network %s (%s): conflicts with network %s (%s): %s",
-					config.ID, config.BridgeName, nwConfig.ID, nwConfig.BridgeName, err.Error())
+					config.ID, config.BridgeName, nwID, nwObj.config.BridgeName, err.Error())
 			}
-		}
-	}
 
-	// Create and set network handler in driver
-	network := &bridgeNetwork{
-		id:         config.ID,
-		endpoints:  make(map[string]*bridgeEndpoint),
-		config:     config,
-		portMapper: portmapper.New(d.config.UserlandProxyPath),
-		driver:     d,
-	}
-
-	d.Lock()
-	d.networks[config.ID] = network
-	d.Unlock()
+			// TODO fcrisciani following is SUPER UGLY is basically a copy paste of the previous DeleteNetwork needs cleanup
```

Comment:
I agree. This is super ugly and lets try to avoid this at all cost. If you are moving the DeleteNetwork content out here to avoid the d.Lock() deadlock, that means its a surgical operation and can lead to long-term maintenance headaches. It would be better to think about other ease-of-maintenance approaches.